### PR TITLE
fix spark-kubernetes-operator compatibility

### LINKED ISSUE #31183: SparkKubernetesSensor: 'None' has no attribute 'metadata'
### Apache Airflow version

2.6.0

### What happened

After upgrading to 2.6.0 version, pipelines with SparkKubernetesOperator -> SparkKubernetesSensor stopped working correctly.

[this PR](https://github.com/apache/airflow/pull/29977) introduces some enhancement into Spark Kubernetes logic, now SparkKubernetesOperator receives the log from spark pods (which is great), but it doesn't monitor the status of a pod, which means if spark application fails - a task in Airflow finishes successfully. 
On the other hand, using previous pipelines (Operator + Sensor) is impossible now, cause SparkKubernetesSensor fails with `jinja2.exceptions.UndefinedError: 'None' has no attribute 'metadata'` as SparkKubernetesOperator is no longer pushing info to xcom. 

### What you think should happen instead

Old pipelines should be compatible with Airflow 2.6.0, even though the log would be retrieved in two places - operator and sensor. 
OR remove the sensor completely and implement all the functionality in the operator (log + status)

### How to reproduce

Create a DAG with two operators

```
t1 = SparkKubernetesOperator(
        kubernetes_conn_id='common/kubernetes_default',
        task_id=f"task-submit",
        namespace="namespace",
        application_file="spark-applications/app.yaml",
        do_xcom_push=True,
        dag=dag,
    )

    t2 = SparkKubernetesSensor(
        task_id=f"task-sensor",
        namespace="namespace",
        application_name=f"{{{{ task_instance.xcom_pull(task_ids='task-submit')['metadata']['name'] }}}}",
        dag=dag,
        attach_log=True,
    )

```

### Operating System

Debian GNU/Linux 10 (buster)

### Versions of Apache Airflow Providers

apache-airflow-providers-amazon==8.0.0
apache-airflow-providers-apache-spark==4.0.1
apache-airflow-providers-celery==3.1.0
apache-airflow-providers-cncf-kubernetes==6.1.0
apache-airflow-providers-common-sql==1.4.0
apache-airflow-providers-docker==3.6.0
apache-airflow-providers-elasticsearch==4.4.0
apache-airflow-providers-ftp==3.3.1
apache-airflow-providers-google==10.0.0
apache-airflow-providers-grpc==3.1.0
apache-airflow-providers-hashicorp==3.3.1
apache-airflow-providers-http==4.3.0
apache-airflow-providers-imap==3.1.1
apache-airflow-providers-microsoft-azure==6.0.0
apache-airflow-providers-microsoft-mssql==3.3.2
apache-airflow-providers-microsoft-psrp==2.2.0
apache-airflow-providers-microsoft-winrm==3.1.1
apache-airflow-providers-mysql==5.0.0
apache-airflow-providers-odbc==3.2.1
apache-airflow-providers-oracle==3.6.0
apache-airflow-providers-postgres==5.4.0
apache-airflow-providers-redis==3.1.0
apache-airflow-providers-sendgrid==3.1.0
apache-airflow-providers-sftp==4.2.4
apache-airflow-providers-slack==7.2.0
apache-airflow-providers-snowflake==4.0.5
apache-airflow-providers-sqlite==3.3.2
apache-airflow-providers-ssh==3.6.0
apache-airflow-providers-telegram==4.0.0

### Deployment

Other 3rd-party Helm chart

### Deployment details

_No response_

### Anything else

_No response_

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by blcksrx**
This PR would fix the problems mentioned in the issue.

- with new args `watch: bool = False` this operator executes as same as the old version and the compatibility issue would solves
- The job status covers by the k8s events in case of success or fail
- tests for these situation

closes: #31183

---
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

**Comment by blcksrx** [comment]
@hanna-liashchuk @Owen-CH-Leung @matheus-rossi @adminidever
Since you were involved with the related issue, please don't hesitate to review this PR

**Comment by jherrmannNetfonds** [comment]
Hi, thanks for the PR.
I am testing this in production right now with airflow 2.6.1 running on Kubernetes with KubernetesExecutor. I encountered this error today:

```
[2023-06-16, 09:44:04 CEST] {taskinstance.py:1824} ERROR - Task failed with exception
Traceback (most recent call last):
  File "/home/airflow/.local/lib/python3.10/site-packages/urllib3/response.py", line 761, in _update_chunk_length
    self.chunk_left = int(line, 16)
ValueError: invalid literal for int() with base 16: b''
During handling of the above exception, another exception occurred:
Traceback (most recent call last):
  File "/home/airflow/.local/lib/python3.10/site-packages/urllib3/response.py", line 444, in _error_catcher
    yield
  File "/home/airflow/.local/lib/python3.10/site-packages/urllib3/response.py", line 828, in read_chunked
    self._update_chunk_length()
  File "/home/airflow/.local/lib/python3.10/site-packages/urllib3/response.py", line 765, in _update_chunk_length
    raise InvalidChunkLength(self, line)
urllib3.exceptions.InvalidChunkLength: InvalidChunkLength(got length b'', 0 bytes read)
During handling of the above exception, another exception occurred:
Traceback (most recent call last):
  File "/opt/airflow/dags/repo/packages/local_copy_of_this_pr/spark_kubernetes.py", line 122, in execute
    for line in pod_log_stream:
  File "/home/airflow/.local/lib/python3.10/site-packages/kubernetes/watch/watch.py", line 165, in stream
    for line in iter_resp_lines(resp):
  File "/home/airflow/.local/lib/python3.10/site-packages/kubernetes/watch/watch.py", line 56, in iter_resp_lines
    for seg in resp.stream(amt=None, decode_content=False):
  File "/home/airflow/.local/lib/python3.10/site-packages/urllib3/response.py", line 624, in stream
    for line in self.read_chunked(amt, decode_content=decode_content):
  File "/home/airflow/.local/lib/python3.10/site-packages/urllib3/response.py", line 816, in read_chunked
    with self._error_catcher():
  File "/usr/local/lib/python3.10/contextlib.py", line 153, in __exit__
    self.gen.throw(typ, value, traceback)
  File "/home/airflow/.local/lib/python3.10/site-packages/urllib3/response.py", line 461, in _error_catcher
    raise ProtocolError("Connection broken: %r" % e, e)
```
This fails the task, but the spark application is still running and succeeding. Maybe some of these types of errors should be catched instead of letting the pod fail. What do you think?

**Comment by blcksrx** [comment]
> @jherrmannNetfonds Good catch buddy. Since with `watch=True` this operator is responsible for the execution, logs and states of the task, I believe in that case the pod/job should be killed, that's why I overwrited the `on_kill` method to kill it. But apparently the `on_kill` didn't work.

@potiuk 
I believe this behaviour that `on_kill` method did not invokes is because we only catch `AirflowTaskTimeout`
https://github.com/apache/airflow/blob/a60429eadfffb5fb0f867c220a6cecf628692dcf/airflow/models/taskinstance.py#L1709-L1718

**Comment by potiuk** [comment]
> > @jherrmannNetfonds Good catch buddy. Since with `watch=True` this operator is responsible for the execution, logs and states of the task, I believe in that case the pod/job should be killed, that's why I overwrited the `on_kill` method to kill it. But apparently the `on_kill` didn't work.
> 
> @potiuk I believe this behaviour that `on_kill` method did not invoke is because we only catch `AirflowTaskTimeout`
> 
> https://github.com/apache/airflow/blob/a60429eadfffb5fb0f867c220a6cecf628692dcf/airflow/models/taskinstance.py#L1709-L1718

We run on_kill in local_task_runner:

```
        finally:
            self.on_kill()
```

**Comment by potiuk** [comment]
> > > @jherrmannNetfonds Good catch buddy. Since with `watch=True` this operator is responsible for the execution, logs and states of the task, I believe in that case the pod/job should be killed, that's why I overwrited the `on_kill` method to kill it. But apparently the `on_kill` didn't work.
> > 
> > 
> > @potiuk I believe this behaviour that `on_kill` method did not invoke is because we only catch `AirflowTaskTimeout`
> > https://github.com/apache/airflow/blob/a60429eadfffb5fb0f867c220a6cecf628692dcf/airflow/models/taskinstance.py#L1709-L1718
> 
> We run on_kill in local_task_runner:
> 
> ```
>         finally:
>             self.on_kill()
> ```

**Comment by blcksrx** [comment]
@potiuk Why closing PR?
With a try catch it would be done

**Comment by jherrmannNetfonds** [comment]
Please consider reopening the PR, since 2.6.x came with braking changes for the `SparkKubernetesOperator` compared to 2.5.x which would be "fixed" in this PR. With 2.6.x the `SparkKubernetesOperator` is basically unusable, since it will fail to often in a pure k8 environment compared to the sensor approach in before 2.6.0

**Comment by potiuk** [comment]
Sorry, Sorry, it was **PURELY** accidental :). I usually write briefly why I am closing stuff (sometimes maybe in a harsh way but I usually tell why ;)

**Comment by jherrmannNetfonds** [comment]
thx for reopening.
@blcksrx fyi
I am testing my local version from line 123 with
```
try:
    self.log.info(line)
except HTTPError as e:
    self.log.warn(e)
```
Probably cleaner to kill everything (task + sparkapp) on an error, but on the other side, sparkapps are mostly long running jobs processing big amounts of data. So the question is, what is more intuitive / less confusing / less annoying for the user.

File: pyproject.toml [hunk 1]
```diff
@@ -75,7 +75,7 @@ extend-ignore = [
 # * Disable `flaky` plugin for pytest. This plugin conflicts with `rerunfailures` because provide same marker.
 # * Disable `nose` builtin plugin for pytest. This feature deprecated in 7.2 and will be removed in pytest>=8
 # * And we focus on use native pytest capabilities rather than adopt another frameworks.
-addopts = "-rasl --verbosity=2 -p no:flaky -p no:nose --asyncio-mode=strict"
```
Comment (potiuk) [inline pyproject.toml]: Why changing that ?

Comment (blcksrx) [reply pyproject.toml]: Oh sorry, that was a clumsy action 😁

**Review by potiuk — APPROVED** [review]
I do not know much about the Spark/K8S interface, but the change looks legit.

@jedcunningham - would you want to take a look as well?

**Comment by sudohainguyen** [comment]
hey @blcksrx I just got an issue when my SparkApp completed but never reached `SparkApplicationCompleted` reason, the airflow task is stuck into infinite loop until it gets 410 error

Have you encountered the same ?

**Pull request merged**